### DISCUSSION TITLE
Should there be an list Type for columns, tables etc.?

### DISCUSSION THREAD

**Discussion by Gerhardsa0** [Safe-DS/.github#4] [Library Discussion]
Currently, I am working on a new time series class in the standard library. When working with time series, the data often needs to be split into "windows" so the model can better learn the dependencies between the data points. Right now, the safe-ds std lib does not allow np.ndarrays as a datatype. I think it is good to allow this datatype to support "windows" easier and also be efficient. (Lists are also not implemented)
Should there be an option to work with lists as a datatype, or is this something that should not be allowed?
If we don't allow it, then the workaround will be intense.
@lars-reimann

**Comment by lars-reimann** [comment]
I'd consider `ndarrays` (or a custom wrapper at a similar abstraction level) to be too low-level for our library. Some thoughts/questions:

* How about adding a `Window` class that can offer a high-level API to work with window data?  This would be similar to the `Row`/`Column` classes that we offer for tabular data.
* What operations would you like to do on a window?
* What means would you offer to move between a time series and one or more windows?

**Reply by Gerhardsa0** [reply]
- I think this is a great idea, something like a window class, which allows multiple data points in one row. So we can do something like this:
 Col = [window, window, window], so each row in Col  can take multiple data points. This would be more similiar to a column_type. 
 
-  Then something like a windowed_column would be similair to a row and column, but I am unsure if that is needed/wanted, because right now I cant really think about special window operations(most likely they will just be used as Input into a ML-Model)

- Often the windowsize needs to be figured out by adjusting the windowsize, so maybe multiple windows would be needed for that usecase

When we add more functionalitys to the timeseries its likely something like a extra windowed_column would be needed, but I am not sure.

thanks for the feedback!

**Reply by lars-reimann** [reply]
> windowed_column would be similair to a row and column, but I am unsure if that is needed/wanted

Currently, I don't see the need for that either. `TimeSeries` and `Window` should suffice for a start. If we later find use cases for a windowed column, we can add that.
